### PR TITLE
Shared task options and docker build cleanup

### PR DESCRIPTION
**Pull request by lancetipton**
**Ticket**: [ZEN-463](https://jira.simpleviewtools.com/browse/ZEN-463)

[Semver-Status] minor

> These changes are dependent on [this PR](https://github.com/simpleviewinc/keg-hub/pull/92)
> Please review those changes first. Once they are merged, I will update the PR to point to the develop branch

## Context

* These changes are part of a bigger ticket to update how docker images are managed. 
* These changes were extracted to to allow for a smaller PR. Because of this; some functionality may not be used in this PR
* To allow better management of docker images, we need to improve how images are tagged

## Goal

* Allow tasks to share common options
  * Common options only need to be updated in one location
  * They can then be reused for tasks that use that option
* Improve docker build functionality
  * Add more build options
* Added better tagging functionality for images

## Updates

* `repos/keg-cli/src/constants/tagMap.js`
  * Constants used when parse task options into docker image tags
* `repos/keg-cli/src/tasks/*`
  * Updates some tasks to share common options between them
* `repos/keg-cli/src/utils/docker/buildDockerCmd.js`
  * Refactored the docker build command to work with image tag updates
* `repos/keg-cli/src/utils/docker/getDockerCmdArgs.js`
  * Renamed from `getDockerArgs` to better specify what the function does
* `repos/keg-cli/src/utils/docker/tags/*`
  * Added a number of build tag helpers to parse images tags dynamically
* `repos/keg-cli/src/utils/task/options/*`
  * Shared task options
  * This allows updating the options in one location, which then updates all tasks that use them

## Testing

**Setup**
* Pull this PR => `keg cli && keg pr 91`
* Install the latest deps => `yarn install`
* Clear out all your old docker images and cache
  * `keg d clean -f`
  * Ensure the `keg-base` image is removed
    * Run `keg di rm`
    * Then select the base image if it still exists

**Test 1**
* Run the command `keg base build`
  * This should build the base image without issue
* Run the command `keg components build`
  * This should build the `keg-components` image without issue
* Run the command `keg evf build`
  * This should build the `tap` image without issue

**Test 2**
* Remove all the image images built in **Test 1**
* Run the command `keg components start`
  * This should build the `keg-base` && `keg-components` images without issue
  * It should then start the `keg-components` container
  * Ensure keg-components runs as expected
* Run the command `keg evf start`
  * This should build only the `tap` image without issue, because the base image was built in the last step
  * It should then start the `events force tap` container
  * Ensure the tap events force runs as expected

**Test 3**
* Run `keg evf pack`
  * This tests to ensure the `docker package` workflow still works properly
  * Ensure the package is created and pushed as expected

**Test 4**
* Run `keg evf pack run package=docker.pkg.github.com/simpleviewinc/keg-packages/tap:zen-468-pressable-update`
  * This tests to ensure the `docker provider pull` task still works as expected
  * Ensure the package runs as expected

**Test 5**
* Run the command `keg cli test`
  * Ensure all cli tests pass

### CODE REVIEW

**Review by mpcarolin — CHANGES_REQUESTED** [review]
When running `keg evf start` for `Step 2`, I encountered this error:

image

**Comment by lancetipton** [comment]
@mpcarolin 
* I went through the steps exactly 1 by 1 and everything worked as expected.
  * I actually went through the steps twice
* Here's a shot of running the  TAP =>  http://snpy.in/bhkhyb
  * Comparing the build output between the 2 images, there's a few differences
    * Did you follow the first bullet point of **Test 2**?
    * From the image  you posted, you have a number of `cached` steps
    * `cached` means the image was already built once before?

**Comment by mpcarolin** [comment]
> @mpcarolin
> 
>     * I went through the steps exactly 1 by 1 and everything worked as expected.
>       
>       * I actually went through the steps twice
> 
>     * Here's a shot of running the  TAP =>  http://snpy.in/bhkhyb
>       
>       * Comparing the build output between the 2 images, there's a few differences
>         
>         * Did you follow the first bullet point of **Test 2**?
>         * From the image  you posted, you have a number of `cached` steps
>         * `cached` means the image was already built once before?
> 
> 
> image

So I went through it again, and I'm still encountering the same thing. I followed the steps exactly. I also tried deleting all the images and starting just `evf`, and I get the same thing:

image

Maybe we can hop on a call.

File: repos/keg-cli/src/utils/getters/getTagVarMap.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * <br/>Also maps each alias to a key name, which is then used to call the correct method. Used in tagFromVariable helper
+ * @type Object
+ */
+reduceObj(tagMap, (key, value, mapped) => {
+  value.alias.map(alias => aliasToTag[alias] = key)
+
+  value.combine.map(otherTag => {
+    if(!tagMap[otherTag] || !tagMap[otherTag].alias || !tagMap[otherTag].alias.length) return
+
+    const otherAlias = tagMap[otherTag].alias
```
Comment (daniel112) [inline repos/keg-cli/src/utils/getters/getTagVarMap.js:18]: i dont think this is being used?

**Review by daniel112 — APPROVED** [review]
All tests passed.
left a few minor comments but it's not blocking.
you can ignore them if those files are going to be updated in the next PR anyways

**Review by mpcarolin — APPROVED** [review]
Passed all testing!

File: repos/keg-cli/src/utils/task/options/pushOptions.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
+const { fromImage, tagVariable } = require('./singleOptions')
```
[the remark this answers is not in the document]
Comment (lancetipton) [reply repos/keg-cli/src/utils/task/options/pushOptions.js:1]: Updated in PR #92

File: repos/keg-cli/src/utils/docker/tags/tagHelpers.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+/**
+ * Gets a git repos current commit hash or branch name based on the containerContext and location
+ * @type function
+ * @param {Object} containerContext - Metadata about the image to be built ( from buildContainerContext helper )
+ * @param {Array} params - Options passed from the command line
+ *
+ * @returns {Promise[string]} - Current commit hash or branch name 
+ */
+const getRepoGitTag = async ({ containerContext, params }, method) => {
```
[the remark this answers is not in the document]
Comment (lancetipton) [reply repos/keg-cli/src/utils/docker/tags/tagHelpers.js:82]: Updated in PR #92

File: repos/keg-cli/src/utils/docker/tags/tagFromVersion.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ * Gets the version tag from the passed in option, the package.json or context
+ * @type function
+ * Formats found tags option into docker format
+ * @param {Array} params - Options passed from the command line
+ * @param {string} env - Current environment the command is being run in
+ * @param {Object} args - Arguments passed to the task
+ *
+ * @returns {string} - Formatted string of tags for docker
+ */
+const tagFromVersion = async (params, args) => {
```
[the remark this answers is not in the document]
Comment (lancetipton) [reply repos/keg-cli/src/utils/docker/tags/tagFromVersion.js:14]: Updated in PR #92

File: repos/keg-cli/src/utils/getters/getTagVarMap.js (hunk 1 above)
Comment (lancetipton) [reply repos/keg-cli/src/utils/getters/getTagVarMap.js:18]: It's not right now, but will be in the Image Pull tasks PR #93

**Pull request merged**